packages/react-aria-components/src/Autocomplete.tsx

```diff
@@ -0,0 +1,195 @@
+
+  return (
+    <CollectionBuilder content={content}>
+      {collection => <AutocompleteInner props={props} collection={collection} autocompleteRef={ref} />}
+    </CollectionBuilder>
+  );
+}
+
+interface AutocompleteInnerProps<T extends object> {
+  props: AutocompleteProps<T>,
+  collection: Collection<Node<T>>,
+  autocompleteRef: RefObject<HTMLDivElement | null>
+}
+
+function AutocompleteInner<T extends object>({props, collection, autocompleteRef: ref}: AutocompleteInnerProps<T>) {
+  let {
+    name,
+    formValue = 'key',
+    allowsCustomValue
+  } = props;
+  if (allowsCustomValue) {
+    formValue = 'text';
+  }
+
+  let {validationBehavior: formValidationBehavior} = useSlottedContext(FormContext) || {};
+  let validationBehavior = props.validationBehavior ?? formValidationBehavior ?? 'native';
+  let {contains} = useFilter({sensitivity: 'base'});
+  let state = useAutocompleteState({
+    defaultFilter: props.defaultFilter || contains,
+    ...props,
+    // If props.items isn't provided, rely on collection filtering (aka menu.items is provided or defaultItems provided to Autocomplete)
+    items: props.items,
+    children: undefined,
+    collection,
+    validationBehavior
+  });
+
+  let inputRef = useRef<HTMLInputElement>(null);
+  let menuRef = useRef<HTMLDivElement>(null);
+  let [labelRef, label] = useSlot();
+
+  let {
+    inputProps,
+    menuProps,
+    labelProps,
+    descriptionProps,
+    errorMessageProps,
+    ...validation
+  } = useAutocomplete({
+    ...removeDataAttributes(props),
+    label,
+    inputRef,
+    menuRef,
+    name: formValue === 'text' ? name : undefined,
+    validationBehavior
+  }, state);
+
+  // Only expose a subset of state to renderProps function to avoid infinite render loop
+  let renderPropsState = useMemo(() => ({
+    isDisabled: props.isDisabled || false,
+    isInvalid: validation.isInvalid || false,
+    isRequired: props.isRequired || false
+  }), [props.isDisabled, validation.isInvalid, props.isRequired]);
+
+  let renderProps = useRenderProps({
+    ...props,
+    values: renderPropsState,
+    defaultClassName: 'react-aria-Autocomplete'
+  });
+
+  let DOMProps = filterDOMProps(props);
+  delete DOMProps.id;
+
+  return (
+    <Provider
+      values={[
+        [AutocompleteStateContext, state],
+        [LabelContext, {...labelProps, ref: labelRef}],
+        [InputContext, {...inputProps, ref: inputRef}],
+        [MenuContext, {...menuProps, ref: menuRef}],
+        // TODO: this would need to match the state type of whatever child component the autocomplete
+        // is filtering against... Ideally we'd somehow have the child component communicate its state upwards, upon which we we would filter it from here
+        // and send it back down but that feels circular. However we need a single SelectionManager to be used by the autocomplete and filtered collection's hooks
+        // so that the concepts of "selectedKey"/"focused"
+        // @ts-ignore
```

Comment:
Note: https://github.com/adobe/react-spectrum/tree/submenu_inautocomplete has some work that fixes this by having useAutocompleteState use useTreeState instead